File: apps/dsa-web/src/api/adminLogs.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface ExecutionLogEvent {
  id: number;
  eventAt?: string | null;
  level?: string | null;
  phase: string;
  category?: string | null;
  eventName?: string | null;
  step?: string | null;
  action?: string | null;
  outcome?: string | null;
  reason?: string | null;
  target?: string | null;
  status: string;
  truthLevel: string;
  message?: string | null;
  errorCode?: string | null;
  detail?: Record<string, unknown>;
}

export interface ExecutionLogSessionSummary {
  sessionId: string;
  taskId?: string | null;
  queryId?: string | null;
  analysisHistoryId?: number | null;
  code?: string | null;
  name?: string | null;
  overallStatus: string;
  truthLevel: string;
  startedAt?: string | null;
  endedAt?: string | null;
  summary?: Record<string, unknown>;
  readableSummary?: {
    actorUserId?: string | null;
    actorUsername?: string | null;
    actorDisplay?: string | null;
    actorRole?: string | null;
    sessionKind?: string | null;
    subsystem?: string | null;
    actionName?: string | null;
    destructive?: boolean;
    finalAiModel?: string | null;
    aiAttemptsCount?: number;
    aiFallbackUsed?: boolean;
    finalMarketSource?: string | null;
    finalFundamentalSource?: string | null;
    finalNewsSource?: string | null;
    finalSentimentSource?: string | null;
    dataFallbackUsed?: boolean;
    notificationClassification?: string | null;
    topFailureReason?: string | null;
    scannerRunId?: number | null;
    scannerMarket?: string | null;
    scannerProfile?: string | null;
    scannerProfileLabel?: string | null;
    scannerShortlistCount?: number | null;
    scannerFallbackCount?: number | null;
    scannerProviderFailureCount?: number | null;
    scannerProvidersUsed?: string[];
    scannerCoverageSummary?: string | null;
    summaryParagraph?: string | null;
    status?: string;
    operationCategory?: string | null;
    operationType?: string | null;
    operationTarget?: string | null;
    operationStatus?: string | null;
    keyMetric?: string | null;
    logLevel?: string | null;
    logCategory?: string | null;
    eventName?: string | null;
    eventMessage?: string | null;
    requestId?: string | null;
    source?: string | null;
  };
}

export interface ExecutionLogSessionDetail extends ExecutionLogSessionSummary {
  events: ExecutionLogEvent[];
  operationDetail?: {
    operationCategory?: string | null;
    operationType?: string | null;
    target?: string | null;
    status?: string | null;
    keyMetric?: string | null;
    aiCalls?: Array<Record<string, unknown>>;
    dataSourceCalls?: Array<Record<string, unknown>>;
    systemFallbacks?: Array<string | Record<string, unknown>>;
    systemOperation?: Record<string, unknown>;
    finalResult?: string | null;
    timeline?: Array<Record<string, unknown>>;
    diagnostics?: Array<Record<string, unknown>>;
  };
}

export interface ExecutionLogSessionListResponse {
  total: number;
  items: ExecutionLogSessionSummary[];
  summary?: {
    errorCount?: number;
    warningCount?: number;
    dataSourceFailureCount?: number;
    slowRequestCount?: number;
    latestCriticalAt?: string | null;
  };
}

export interface ExecutionStep {
  name: string;
  label: string;
  provider?: string | null;
  apiPath?: string | null;
  status: string;
  startedAt?: string | null;
  finishedAt?: string | null;
  durationMs?: number | null;
  errorType?: string | null;
  errorMessage?: string | null;
  recordId?: string | null;
  metadata?: Record<string, unknown>;
}

export interface BusinessEvent {
  id: string;
  event: string;
  category: string;
  status: string;
  summary: string;
  symbol?: string | null;
  market?: string | null;
  analysisType?: string | null;
  userId?: string | null;
  requestId?: string | null;
  recordId?: string | null;
  startedAt?: string | null;
  finishedAt?: string | null;
  durationMs?: number | null;
  stepCount: number;
  successStepCount: number;
  failedStepCount: number;
  metadata?: Record<string, unknown>;
}

export interface BusinessEventDetail extends BusinessEvent {
  steps: ExecutionStep[];
}

export interface BusinessEventListResponse {
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
  items: BusinessEvent[];
}

function normalizeSessionSummary(payload: Record<string, unknown>): ExecutionLogSessionSummary {
  const normalized = toCamelCase<ExecutionLogSessionSummary>(payload);
  return {
    ...normalized,
    readableSummary: normalized.readableSummary && typeof normalized.readableSummary === 'object'
      ? normalized.readableSummary
      : {},
  };
}

function normalizeSessionDetail(payload: Record<string, unknown>): ExecutionLogSessionDetail {
  const normalized = toCamelCase<ExecutionLogSessionDetail>(payload);
  return {
    ...normalized,
    readableSummary: normalized.readableSummary && typeof normalized.readableSummary === 'object'
      ? normalized.readableSummary
      : {},
    events: Array.isArray(normalized.events) ? normalized.events : [],
    operationDetail: normalized.operationDetail && typeof normalized.operationDetail === 'object'
      ? normalized.operationDetail
      : {},
  };
}

export const adminLogsApi = {
  listBusinessEvents: async (
    params: {
      category?: string;
      symbol?: string;
      status?: string;
      query?: string;
      since?: string;
      limit?: number;
      offset?: number;
    },
  ): Promise<BusinessEventListResponse> => {
    const response = await apiClient.get<Record<string, unknown>>(
      '/api/v1/admin/logs',
      { params },
    );
    const normalized = toCamelCase<BusinessEventListResponse>(response.data);
    return {
      total: Number(normalized.total || 0),
      limit: Number(normalized.limit || params.limit || 50),
      offset: Number(normalized.offset || params.offset || 0),
      hasMore: Boolean(normalized.hasMore),
      items: Array.isArray(normalized.items)
        ? normalized.items.map((item) => toCamelCase<BusinessEvent>(item as unknown as Record<string, unknown>))
        : [],
    };
  },

  getBusinessEventDetail: async (eventId: string): Promise<BusinessEventDetail> => {
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/admin/logs/${encodeURIComponent(eventId)}`,
    );
    const normalized = toCamelCase<BusinessEventDetail>(response.data);
    return {
      ...normalized,
      steps: Array.isArray(normalized.steps) ? normalized.steps : [],
    };
  },

  listSessions: async (
    params: {
      taskId?: string;
      stock?: string;
      status?: string;
      minLevel?: string;
      level?: string;
      category?: string;
      query?: string;
      provider?: string;
      model?: string;
      channel?: string;
      since?: string;
      limit?: number;
      offset?: number;
    },
  ): Promise<ExecutionLogSessionListResponse> => {
    const requestParams = {
      ...params,
      min_level: params.minLevel,
      task_id: params.taskId,
    };
    delete requestParams.minLevel;
    delete requestParams.taskId;
    const response = await apiClient.get<Record<string, unknown>>(
      '/api/v1/admin/logs/sessions',
      {
        params: requestParams,
      },
    );
    const normalized = toCamelCase<ExecutionLogSessionListResponse>(response.data);
    return {
      total: Number(normalized.total || 0),
      summary: normalized.summary,
      items: Array.isArray(normalized.items)
        ? normalized.items.map((item) => normalizeSessionSummary(item as unknown as Record<string, unknown>))
        : [],
    };
  },

  getSessionDetail: async (sessionId: string): Promise<ExecutionLogSessionDetail> => {
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/admin/logs/sessions/${encodeURIComponent(sessionId)}`,
    );
    return normalizeSessionDetail(response.data);
  },
};

File: api/v1/schemas/admin_logs.py
# -*- coding: utf-8 -*-
"""Schemas for admin execution logs."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class ExecutionLogEventModel(BaseModel):
    id: int
    event_at: Optional[str] = None
    level: Optional[str] = None
    phase: str
    category: Optional[str] = None
    event_name: Optional[str] = None
    step: Optional[str] = None
    action: Optional[str] = None
    outcome: Optional[str] = None
    reason: Optional[str] = None
    target: Optional[str] = None
    status: str
    truth_level: str
    message: Optional[str] = None
    error_code: Optional[str] = None
    detail: Dict[str, Any] = Field(default_factory=dict)


class ExecutionLogSessionSummaryModel(BaseModel):
    session_id: str
    task_id: Optional[str] = None
    query_id: Optional[str] = None
    analysis_history_id: Optional[int] = None
    code: Optional[str] = None
    name: Optional[str] = None
    overall_status: str
    truth_level: str
    started_at: Optional[str] = None
    ended_at: Optional[str] = None
    summary: Dict[str, Any] = Field(default_factory=dict)
    readable_summary: Dict[str, Any] = Field(default_factory=dict)


class ExecutionLogSessionDetailModel(ExecutionLogSessionSummaryModel):
    events: List[ExecutionLogEventModel] = Field(default_factory=list)
    operation_detail: Dict[str, Any] = Field(default_factory=dict)


class ExecutionLogSummaryModel(BaseModel):
    error_count: int = 0
    warning_count: int = 0
    data_source_failure_count: int = 0
    slow_request_count: int = 0
    latest_critical_at: Optional[str] = None


class ExecutionLogSessionListResponse(BaseModel):
    total: int
    items: List[ExecutionLogSessionSummaryModel] = Field(default_factory=list)
    summary: ExecutionLogSummaryModel = Field(default_factory=ExecutionLogSummaryModel)


class ExecutionStepModel(BaseModel):
    name: str
    label: str
    provider: Optional[str] = None
    apiPath: Optional[str] = None
    status: str
    startedAt: Optional[str] = None
    finishedAt: Optional[str] = None
    durationMs: Optional[float] = None
    errorType: Optional[str] = None
    errorMessage: Optional[str] = None
    recordId: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class BusinessEventModel(BaseModel):
    id: str
    event: str
    category: str
    status: str
    summary: str
    symbol: Optional[str] = None
    market: Optional[str] = None
    analysisType: Optional[str] = None
    userId: Optional[str] = None
    requestId: Optional[str] = None
    recordId: Optional[str] = None
    startedAt: Optional[str] = None
    finishedAt: Optional[str] = None
    durationMs: Optional[float] = None
    stepCount: int = 0
    successStepCount: int = 0
    failedStepCount: int = 0
    metadata: Dict[str, Any] = Field(default_factory=dict)


class BusinessEventDetailModel(BusinessEventModel):
    steps: List[ExecutionStepModel] = Field(default_factory=list)


class BusinessEventListResponse(BaseModel):
    items: List[BusinessEventModel] = Field(default_factory=list)
    total: int
    limit: int
    offset: int
    hasMore: bool = False
